modules/mockk-agent-android/build.gradle.kts

```diff
@@ -12,6 +12,8 @@ description = "Android instrumented testing MockK inline mocking agent"
 val mavenName: String by extra("MockK Android Agent")
 val mavenDescription: String by extra("${project.description}")
 
+val dispatcherJarResPath: Provider<Directory> = layout.buildDirectory.dir("generated/dispatcher-jar")
+
```

Comment:
KTS is basically executed in the order described.
In the previous description, this value is `null` at the time `srcDirs` is set. (Kotlin spec)
Therefore, writing this value before `srcDirs` solves this problem.

Reply:
I'd be surprised if this makes a difference, because `AndroidSourceDirectorySet.srcDirs` is documented as accepting providers...

Perhaps what's going wrong is that the task `preBuild` isn't being triggered. I think what will probably work better is if the task output is used instead.

```kotlin
    sourceSets {
        named("main").configure {
            resources {
                srcDirs(packageDispatcherJar.map { it.destinationDirectory })
            }
        }
    }
```

Reply:
This problem is caused by the initialization order of Kotlin Script.
Decompile KTS and you will see why it occurs.

Decompiled code when `dispatcherJarResPath` is written before `android {}`.
```java
public class Build_gradle extends KotlinBuildScript {
   // ...
   @NotNull
   private final Provider dispatcherJarResPath;

   @NotNull
   public final Provider getDispatcherJarResPath() {
      return this.dispatcherJarResPath;
   }
   // ...

   public Build_gradle(KotlinScriptHost var1) {
      // ...
      // val dispatcherJarResPath: Provider<Directory> = layout.buildDirectory.dir("generated/dispatcher-jar")
      Provider var10002 = ((Build_gradle)this).getLayout().getBuildDirectory().dir("generated/dispatcher-jar");
      Intrinsics.checkNotNullExpressionValue(var10002, "layout.buildDirectory.di…enerated/dispatcher-jar\")");
      this.dispatcherJarResPath = var10002;
      // android {
      Accessors7y9hz5zm5pocrb21isjhik2uhKt.android((Project)this, (Action)(new Action() {
            // ...
            // sourceSets {
            $this$android.sourceSets((Function1)(new Function1() {
                  // ...
                  // named("main").configure {
                  $this$sourceSets.named("main").configure((Action)(new Action() {
                        // ...
                        // resources {
                        $this$configure.resources((Function1)(new Function1() {
                           // ...
                           public final void invoke(@NotNull AndroidSourceDirectorySet $this$resources) {
                              Intrinsics.checkNotNullParameter($this$resources, "$receiver");
                              // srcDirs(dispatcherJarResPath)
                              $this$resources.srcDirs(new Object[]{Build_gradle.this.getDispatcherJarResPath()});
                           }
                        }));
                     }
                  }));
               }
            }));
            $this$android.defaultConfig((Function1)null.INSTANCE);
         }
      }));
```

Decompiled code when `val dispatcherJarResPath` is written after `android {}`.

```java
   public Build_gradle(KotlinScriptHost var1) {
      // ...
      // android {
      Accessors7y9hz5zm5pocrb21isjhik2uhKt.android((Project)this, (Action)(new Action() {
            // ...
            // sourceSets {
            $this$android.sourceSets((Function1)(new Function1() {
                  // ...
                  // named("main").configure {
                  $this$sourceSets.named("main").configure((Action)(new Action() {
                        // ...
                        // resources {
                        $this$configure.resources((Function1)(new Function1() {
                           // ...
                           public final void invoke(@NotNull AndroidSourceDirectorySet $this$resources) {
                              Intrinsics.checkNotNullParameter($this$resources, "$receiver");
                              // srcDirs(dispatcherJarResPath)
                              // ⚠️ ⚠️ getDispatcherJarResPath() returns null!
                              $this$resources.srcDirs(new Object[]{Build_gradle.this.getDispatcherJarResPath()});
                           }
                        }));
                     }
                  }));
               }
            }));
            $this$android.defaultConfig((Function1)null.INSTANCE);
         }
      }));
      // ...
      // val dispatcherJarResPath: Provider<Directory> = layout.buildDirectory.dir("generated/dispatcher-jar")
      Provider var10002 = ((Build_gradle)this).getLayout().getBuildDirectory().dir("generated/dispatcher-jar");
      Intrinsics.checkNotNullExpressionValue(var10002, "layout.buildDirectory.di…enerated/dispatcher-jar\")");
      this.dispatcherJarResPath = var10002;
```

Reply:
Alright, I understand now, thanks for explaining. I thought that the value of the provider was null, but I see now it was the provider variable itself.